docs/dev/sql-nested-function-where-clause.md

```diff
@@ -0,0 +1,397 @@
+
+## Table of Contents
+1. [Overview](#1-overview)
+2. [Syntax](#11-syntax)
+3. [Changes To Core](#12-changes-to-core)
+4. [Example Data And Usage](#13-example-data-and-usage)
+5. [Example Queries](#14-example-queries)
+6. [Architecture Diagrams](#2-architecture-diagrams)
+7. [Sequence Diagram for filter Push Down](#21-sequence-diagram-for-filter-push-down)
+8. [Syntax Option 1 Object Tree to DSL](#22-syntax-option-1-object-tree-to-dsl)
+9. [Syntax Option 2 Object Tree to DSL](#23-syntax-option-2-object-tree-to-dsl)
+10. [Additional Info](#additional-info)
+11. [Release Schedule](#release-schedule)
+
+## 1 Overview
+### 1.1 Syntax
+
+The nested function has two syntax options when used in the `WHERE` clause of an SQL statement. Both syntax options a
+1. `nested(path, condition)`
+2. `nested(field | field, path) OPERATOR LITERAL`
+
+### 1.2 Changes To Core
+- **FilterQueryBuilder:** Added logic to handle `nested` functions in `WHERE` clause as predicate expression and `FunctionExpression`.
+- **LuceneQuery:** Added logic to handle `nested` functions and `FunctionExpression`'s on both sides of `OPERATOR` in `WHERE` clause.
+
+
+### 1.3 Example Data and Usage
+
+A basic example from data mapping to response from SQL plugin. This example is very basic but useful for understanding the `nested` functions use.
+
+**Mapping:**
+```json
+{
+  "mappings": {
+    "properties": {
+      "message": {
+        "type": "nested",
+        "properties": {
+          "info": {
+            "type": "keyword",
+            "index": "true"
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+**Dataset:**
+```json
+{"index":{"_id":"1"}}
+{"message":{"info":"a"}}
+{"index":{"_id":"2"}}
+{"message":{"info":"c"}}
+```
+
+**Query:**
+- `SELECT * FROM nested_objects WHERE nested(message, message.info = 'a');`
+- `SELECT * FROM nested_objects WHERE nested(message.info) = 'a';`
+
+Both queries produce same response from OpenSearch.
+**Response:**
+```json
+{
+  "schema": [
+    {
+      "name": "message",
+      "type": "nested"
+    }
+  ],
+  "datarows": [
+    [
+      [
+        {
+          "info": "a"
+        }
+      ]
+    ]
+  ],
+  "total": 1,
+  "size": 1,
+  "status": 200
+}
+```
+
+### 1.4 Example Queries
+
+A basic nested function in the `SELECT` clause. This example filters the `nested` object `message` and the inner field `info` with the value of `a`.
+- `SELECT * FROM nested_objects WHERE nested(message.info) = 'a' OR nested(message.author) = 'elm';`
+
+The same above query but using syntax option 1. The response from OpenSearch will be the same, but the DSL used in the query will differ.
```

Comment:
The first sentence is a little bit clunky, maybe something like this would sound better?

```suggestion
Using the same query as above but with syntax option 1, the response from OpenSearch will be the same, but the DSL used in the query will differ.
```

Reply:
Fixed in [fafca288aad64c8c4be3ff8bce2e49668bc7a4d2](https://github.com/Bit-Quill/opensearch-project-sql/pull/265/commits/fafca288aad64c8c4be3ff8bce2e49668bc7a4d2).